---
title: "Analysis of glioblastoma (GBM) scRNA-seq samples with scBubbletree"
author: "Simo Kitanovski (simo.kitanovski@uni-due.de)"
output:
  BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{GBM data analysis with scBubbletree}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  markdown: 
    wrap: 72
---

```{r setup, include = FALSE, warning = FALSE}
knitr::opts_chunk$set(comment = FALSE, warning = FALSE, message = FALSE)
```

# Background

Here we analyze a publicly available scRNA-seq dataset with over 45,000
cells from five human glioblastoma (GBM) patients[^1][^2]. Paired
biopsies from these three tissues have been isolated from each patient:
tumor center, peripheral infiltration zone and blood. With this vignette
we want to demonstrate the versatility of `r Biocpkg("scBubbletree")` in
analyzing complex datasets, such as scRNA-seq data from solid tumors.

First, we demonstrate how to extract biological insights from scRNA-seq
data using `r Biocpkg("scBubbletree")` visuals at the very start of any
scRNA-seq data analysis. Second, we compare our results with those
presented in the original publication.

To run this vignette we need to load a few packages:

```{r}
library(scBubbletree)
library(ggplot2)
library(ggtree)
library(patchwork)
```


# Data[^1]

Here we will analyze a publicly available scRNA-seq data[^2] generated
10X Genomics Chromium Single Cell Controller. The raw data (barcode-gene
count matrix) is accompanied by supplementary meta-data for each cell,
including:


-   cell type labels (general and specific)
-   cell phase labels
-   sample name
-   tissue name
-   quality metrics: % of transcripts coming from
    mitochondrial/ribosomal genes


# Data processing

Data processing was performed with the R-package `r CRANpkg("Seurat")`.
Gene expressions were normalized (function `NormalizeData`),
highly-variable genes were identified (function `FindVariableFeatures`),
and the expressions were scaled gene-wise (function `ScaleData`). Next,
principal component analysis (PCA) was performed with the function
`RunPCA` based on the 5,000 most variable genes.

In the data we saw that the first 15 principal components capture most
of the variance in the data, and the proportion of variance explained by
each subsequent principal component was negligible. Thus, we used the
single cell projections (embeddings) in 15-dimensional feature space,
$A^{45,466\times 15}$.

```{r}
# # This script can be used to generate data("GBM_ccl", package="scBubbletree")
# 
# # create directory
# dir.create(path = "case_study/")
# 
# # download the data from:
# https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE197543
#
# extract data: 
# gzip -d GSE197543_UMIsMatrix.txt.gz
# gzip -d GSE197543_colData.txt.gz
# 
# # read.csv
# u <- read.csv(file = "case_study/GSE197543_UMIsMatrix.txt", sep = "\t")
# 
# # create Seurat object from the raw counts and append the meta data to it
# d <- Seurat::CreateSeuratObject(counts = u, project = 'GBM')
# 
# # get meta
# meta <- read.csv(file = "data/CS_Glio/GSE197543_colData.txt", sep = "\t")
# meta_tissue <- do.call(rbind, strsplit(x = rownames(d@meta.data), split = "_"))
# d@meta.data <- cbind(d@meta.data, meta)
# d@meta.data$tissue <- meta_tissue[,4]
# d@meta.data$patient <- meta_tissue[,3]
# d@meta.data$cell_id <- 1:nrow(d@meta.data)
# 
# # normalize data
# d <- NormalizeData(object = d, normalization.method = "LogNormalize")
# 
# d <- FindVariableFeatures(object = d, selection.method="vst", nfeatures=5000)
# 
# d <- ScaleData(object = d, features = VariableFeatures(object = d))
# 
# d <- RunPCA(object = d, npcs = 50, features = VariableFeatures(object = d))
# 
# # Run UMAP and tSNE based on top-15 PCs
# d <- RunUMAP(d, dims = 1:15)
# d <- RunTSNE(d, dims = 1:15)
# 
# # Save file to new folder
# save(d, file = "case_study/d.RData")
# # matrix A: the main input of scBubbletree
# GBM_A <- d@reductions$pca@cell.embeddings[, 1:15]
# 
# save(A, file = "case_study/GBM_A.RData")
# save(d@meta.data, file = "case_study/GBM_meta.RData")
```

Load the processed PCA matrix and the meta data

```{r}
# Load the data
data("GBM_A", package = "scBubbletree")
data("GBM_meta", package = "scBubbletree")
data("GBM_exp", package = "scBubbletree")
```

```{r}
data_r <- get(load(file = "../../../code_for_paper/scBubbletree_application/data/CS_Glio_processed/data_r.RData"))
```


# **Step 1:** determining the clustering resolution $r$ for Louvain clustering

Here we use bootstrapping with `B_[\text{gap}}`=10 iteration to determine
the clustering resolution. 

```{r, eval = F}
data_r <- get_r(B_gap = 10, rs = 10^seq(from = -3, to = 1, by = 0.1),
                x = GBM_A, algorithm = "original", cores = 20)
```

The gap statistic increases rapidly until we reach $k'=13$ communities. 
(left panel of the figure below). After $k'=13$ the Gap starts drops to 
values lower than 4. Next, the gap curve increases slowly to 4.13 at $k'=25$,
i.e. even though we double the number of communities, the Gap only increases 
by 0.13. Next, we see another dip and increase in the Gap to 4.22 at $k'=45$.

The Louvain method was used for clustering, hence, from `get_r` we get a 
mapping between the tested resolutions `r` and the corresponding numbers of 
detected communities `k'` (Figure below, right panel).

The resolution $r=0.158$ ($k'=13$) looks like a good first choice for clustering
of the main cell types in this dataset. If the user is interested in e.g. 
finding rare tumor cells, then we recommend increasing the resolution to 
$r=0.79$ ($k'=25$) or even to $r=2.51$ ($k'=45$).

**Remark**: comparable clustering results were reported in [^1], i.e.
they identified between $k=10$ and $k=22$ clusters of transcriptionally 
similar cells with hierarchical clustering:

> Clustering of cells was performed with hierarchical clustering on the 
> Euclidean distances between cells (with Ward’s criterion to minimize the 
> total variance within each cluster Murtagh and Legendre, 2014; package 
> cluster version 2.1.0). The number of clusters used for following analyses 
> was identified by applying a dynamic tree cut (package dynamicTreeCut, 
> version 1.63–1) (Langfelder et al., 2008), resulting in 10 with argument 
> deepSplit set to 1, or 22 clusters with argument deepSplit set to 2.


```{r, fig.width=7, fig.height=3}
(ggplot(data_r$gap_stats_summary)+
   geom_vline(xintercept = c(13, 25, 45), linetype = "dashed", col = "gray")+
   geom_line(aes(y = gap_mean, x = k), col = "darkgray")+
   geom_point(aes(y = gap_mean, x = k), size = 1, col = "black")+
   geom_errorbar(aes(y = gap_mean, x = k, ymin = L95, ymax = H95), width = 0.1)+
   xlab(label = "k'")+
   ylab(label = "Gap")+
   theme_bw(base_size = 10))|
  (ggplot(data_r$gap_stats_summary)+
     geom_vline(xintercept = c(13, 25, 45), linetype = "dashed", col = "gray")+
     geom_line(aes(x = k, y = r), col = "darkgray")+
     geom_point(aes(x = k, y = r), size = 1, col = "black")+
     ylab(label = "r'")+
     xlab(label = "k'")+
     scale_x_log10()+
     scale_y_log10()+
     theme_bw(base_size = 10)+
     annotation_logticks(base = 10, sides = "bl"))
```


# **Step 2-3:** clustering and hierarchical bubble grouping $\rightarrow$ bubbletree

Next, we do Louvain clustering and we will arrange the bubbles in a
bubbletree (hierarchical dendrogram) using hierarchical clustering with 
average linkage.

To construct the bubbletree we will use $B$=300 bootstrap iterations and 
$N_\text{eff}$=100 cells will be drawn at random without replacement from 
each bubble.

```{r}
# graph-based clustering
btd <- get_bubbletree_graph(x = GBM_A,
                            r = 0.158,
                            algorithm = "original",
                            hclust_method = "average",
                            hclust_distance = "euclidean",
                            cores = 1,
                            B = 300,
                            N_eff = 100)
```

# **Step 4:** bubbletree visualization and annotation

We see two outstandingly large bubbles: 0 and 1. Together these two bubbles 
contain about 50% of the cells in this dataset. The smallest bubble 12 contains 
only 184 cells (0.4% of all cells).

```{r, fig.width=5, fig.height=4}
btd$tree
```


## Annotation of cell **type**, **phase**, **tissue** and **gene expression**

`r Biocpkg("scBubbletree")` can integrate categorical and numeric cell
features with the bubbletree visual. Here, we will use the function
`get_cat_tiles` to generate multiple heatmaps, one for each of the
following categorical features which are provided as part of the metadata:

1.  canonical cell type annotations (panel B)
2.  fine-scale cell type annotations (panel C)
3.  cell phases (panel D)
4.  tissue labels (panel E)
5.  gene expression (panel G)

We will use the input parameter `integrate_vertical = FALSE`. This means
that we will show the composition of each bubble (y-axis) with respect
to the categories (labels; x-axis) of each feature.

To visualize numeric features, such as mean gene expressions in the bubbles,
we will use the function `get_num_tiles`.

```{r}
# cell type annotations
g_c <- get_cat_tiles(btd = btd,
                     f = GBM_meta$CellAnnotation,
                     round_digits = 1,
                     tile_text_size = 2,
                     integrate_vertical = FALSE)

# cell type 'fine' annotations 
g_cf <- get_cat_tiles(btd = btd,
                      f = GBM_meta$CellAnnotationFine,
                      round_digits = 1,
                      tile_text_size = 2,
                      integrate_vertical = FALSE)

# cell cycle phase
g_p <- get_cat_tiles(btd = btd,
                     f = GBM_meta$CellCyclePhase,
                     round_digits = 1,
                     tile_text_size = 2,
                     integrate_vertical = FALSE)

# tissue
g_t <- get_cat_tiles(btd = btd,
                     f = GBM_meta$tissue,
                     round_digits = 1,
                     tile_text_size = 2,
                     integrate_vertical = FALSE)

g_e <- get_num_tiles(btd = btd,
                     fs = GBM_exp,
                     summary_function = "mean",
                     round_digits = 1,
                     tile_text_size = 2)

# assemble multiple numeric heatmaps but with individual legends
# g_e <- lapply(X = 1:ncol(GBM_exp), d = GBM_exp, btd = btd, 
#               FUN = function(x, d, btd) {
#                 q <- get_num_tiles(btd = btd,
#                                      fs = GBM_exp[,x],
#                                      summary_function = "mean",
#                                      round_digits = 1,
#                                      tile_text_size = 2)$plot
#                 return(q+theme(legend.position = "none"))
#               })
# 
# g_e <- wrap_plots(g_e, nrow = 1)
```


```{r, fig.width=14, fig.height=10}
(((btd$tree|g_c$plot|g_cf$plot|g_p$plot|g_t$plot)+
   plot_layout(widths = c(1.5, 1, 1.5, 0.4, 0.4)))/
  ((btd$tree|g_e$plot)+
     plot_layout(widths = c(1.2, 1+1.5+0.4+0.4))))+
  plot_annotation(tag_levels = 'A')
```


## What is the **cell type** composition of the bubbles?

First, >95% of the cells in bubbles 8, 5, 11, 10 and 12 are CD45-.
Second, the remaining bubbles are enriched with immune cells. This is
consistent with the results in [^1]:

> "Using hierarchical clustering, the cells were partitioned into clusters 
> (Figure 2-figure supplement 1A and B) which were then annotated into nine 
> distinct cell types for the immune subset, including two transcriptionally 
> distinct MG subsets (MG_1 and MG_2) and four cell types for the CD45-subset 
> (Figure 2A; Figure 2-figure supplement 1C and D; Supplementary file 3)."

Second, the remaining bubbles are immune cells. Importantly, the bubbletree 
conserves global transcriptional distances between these bubbles and splits 
the bubbles in two branches:

  * lymphocytes: B-cells, T-cells, NK-cells, etc.
  * non-lymphocytes: monocytes, neutrophils, microglia and macrophages

Third, we see two special bubbles 3 and 6:

  * bubble 3: tumor-infiltrating lymphocytes $\rightarrow$ the bubble
    has few PBMCs
  * bubble 6: microglia and macrophages

Global distances between the clusters are **not** adequately conserved in 
the 2D tSNE plot (below is the replicated version of Figure 2A from [^1]). 
Furthermore, the 2D tSNE plot suffers from heavy overplotting. This challenge
occurs despite the fact that the cells are split into three panels for each 
tissue site.

```{r, fig.width=6, fig.height=3.5}
ggplot(data = GBM_meta)+
  facet_wrap(facets = ~tissue, nrow = 1)+
  geom_point(aes(x = tSNE1, y = tSNE2, col = CellAnnotation), size = 0.25)+
  theme_bw()+
  theme(legend.position = "top")+
  guides(colour = guide_legend(override.aes = list(size=2)))+
  scale_color_discrete(name = "tissue")
```


## Are all cells from different samples **evenly mixed** within bubbles?

One important question in scRNA-seq data analysis is: "Is the relative 
abundance of cells from different cell types (or bubbles) similar across 
our samples?"

Here, we have data from five patients. For four of these patients we have 
data from three tissues, and for the last patient from two tissues. In total,
we have 14 samples.

We may assume similar relative abundance in a tissue across patients. This 
assumption may be violated due to both biological (e.g. tumors are heterogenous 
both within and between patients) and also technical biases (varying degree 
sampling depth: loss of rare cell types in low-depth samples). Hence, cell 
type composition analysis is regularly performed as part of scRNA-seq data 
analysis.

In fact, from [^1] we get the following remark about the degree of overlap of 
cells from the different cell types across patients:

> As we observed a good overlap of cells across patients for most of
> the dataset (see Methods; Figure 1-figure supplement 2B-F), we chose 
> not to perform any correction for patient-specific effects

> A quantitative test of overlap of cells across patients was made using 
> the CellMixS package (Lütge et al., 2021) (version 1.12.0) developed to 
> quantify the effectiveness of batch correction methods. We used the cell-
> specific mixing (CMS) score, which highlighted a very good overlap across 
> cells from different patients in the lymphoid compartment and for the 
> monocytes. The myeloid compartment displayed a slightly elevated patient-
> specific structure, while it was most pronounced for the CD45 negative 
> subset.

We tried to replicate this analysis with `r Biocpkg("scBubbletree")` (figure 
below). Our analysis with `get_cat_tiles` visualized the composition of patient 
IDs (such IDs are available for each cell) across different bubbles by setting 
`integrate_vertical = TRUE` (panel B). As before, we also show the patient ID 
composition within bubbles by setting `integrate_vertical = FALSE` (panel C).

Surprisingly, we discovered significant biases and we describe them in the
following.


```{r}
g_p_v <- get_cat_tiles(btd = btd,
                     f = paste0(GBM_meta$tissue, '-', GBM_meta$patient),
                     round_digits = 1,
                     tile_text_size = 2,
                     integrate_vertical = TRUE)

g_p_h <- get_cat_tiles(btd = btd,
                     f = paste0(GBM_meta$tissue, '-', GBM_meta$patient),
                     round_digits = 1,
                     tile_text_size = 2,
                     integrate_vertical = FALSE)
```

```{r, fig.width=11, fig.height=6}
(btd$tree|g_p_v$plot|g_p_h$plot)+plot_annotation(tag_levels = 'A')
```


First, we observed similar cell type composition across PBMC samples (panel B).
There were also exceptions: 

  * patient 616 cells were enriched in bubble 0 (monocytes)
  * patient 617 had high relative abundance of cells from bubble 9 
    (neutrophils). From panel C we observe that about 88% of the bubble 9 cells 
    belong to sample PBMC-617

Second, we saw strong deviations in the cell type composition in the tumor 
center and peripheral infiltration zone samples:
  
  * from panel B we see that about 70% of the sample patient 604 (center) 
    and 10.6% of the sample 604 (periphery) are assigned to bubble 5 (NPC-like 
    tumor cells). From panel C we see that nearly all of the cells from this
    bubble are patient 604-specific. The same can be said about bubbles 11, 10
    and 12, i.e. few cells in these bubbles come from patients other than 604.

In conclusion, scBubbletree visualizes the relative abundance of sample/patient 
labels within and across bubbles as categorical features. The resulting visuals 
(panels B and C above) are rich with quantitative information, which facilitates
rapid detection of compositional biases between the samples.


## Quality control with `r Biocpkg("scBubbletree")`

Finally, nearly every scRNA-seq data analysis begins with quality control (QC).

scBubbletree allows you to explore QC metrics commonly used by the community:
  
  1. the number of unique genes detected in each cell
  2. total number of molecules detected within a cell
  3. the percentage of reads that map to the mitochondrial genome

Furthermore, any cell feature (numeric or categorical) that can be uses for QC.
These can be integrated with the bubbletree to identify bubbles enriched with 
low-quality cells.

```{r}
g_qc <- get_num_violins(btd = btd,
                        f = as.matrix(GBM_meta[, c("PercentMitochondrial",
                                                   "PercentRibosomal",
                                                   "NumberDetectedGenes")]))
```

```{r, fig.width=8, fig.height=5}
(btd$tree|g_qc$plot)+plot_layout(widths = c(1.1, 2))
```

# Summary

`r Biocpkg("scBubbletree")` promotes simple and transparent visual
exploration of scRNA-seq. It is **not a black-box approach** and the
user is encouraged to explore the data with different values of $k$ and
$r$ or custom clustering solutions. Attaching of cell features to the
bubbletree is necessary for biological interpretation of the individual
bubbles and their relationships which are described by the bubbletree.


[^1]: <https://doi.org/10.7554/eLife.92678.2>
[^2]: <https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE197543>


# Session Info

```{r}
sessionInfo()
```
